README.md
<!-- 
    Bradley Honeyman
    Oct 6, 2021
-->

# Bootstrap Typography
### A Flutter package that make creating responsive Bootstrap like text easier. Text or SelectableText 
  <br />

  ## Please Post Questions on StackOverflow, and tag @CatTrain (user:16200950)
  https://stackoverflow.com/

  ## Importing:
  ### YAML:
  ```yaml
  dependencies:
    bootstrap_typography2: ^2.1.0
    
  ```
  ### Dart:
  ```dart
  import 'package:bootstrap_typography2/bootstrap_typography2.dart';
  ```

  ## Example:

  ### Simple Examples:
  ```dart
  P(textData: BSTextParams("Hello World!"));
  ```
  ```dart
  H1(
    textData: BSTextParams(
        "Picture",
        weight: FontWeight.bold,
        textAlignment: TextAlign.center,
      ),
  );
  ```
  ```dart
H2(
    textData: BSTextParams(
      "Custom",
      textSizeSmall: 16,
      textSizeLarge: 82,
    ),
);
  ```
```dart
H3(
  textData: BSTextParams(
    "Custom Breakpoints, and Sizes",
    textSizeSmall: 16,
    textSizeLarge: 32,
    customBreakpoints: (
      context,
      textType, {
      required large,
      required small,
    }) {
      if (MediaQuery.of(context).size.width > 500) {
        return large;
      }
      return small;
    },
  ),
),
```
### Simple Selectable Examples:
```dart
P(selectData: BSSelectParams("H1S"));
```
  ```dart
  H1(
    selectData: BSSelectParams(
        "Picture",
        weight: FontWeight.bold,
        textAlignment: TextAlign.center,
      ),
  );
  ```
  ```dart
H2(
    selectData: BSSelectParams(
      "Custom",
      textSizeSmall: 16,
      textSizeLarge: 82,
    ),
);
  ```
```dart
H3(
  selectData: BSSelectParams(
    "Custom Breakpoints, and Sizes",
    textSizeSmall: 16,
    textSizeLarge: 32,
    customBreakpoints: (
      context,
      textType, {
      required large,
      required small,
    }) {
      if (MediaQuery.of(context).size.width > 500) {
        return large;
      }
      return small;
    },
  ),
),
```

  ## Supports:
  Can't pass both ```textData:``` and ```selectData:```
  <br />

  Must pass one: ```textData:``` or ```selectData:```
  
  ### Regular Text:
  - p = ```P(textData: BSTextParams),```
  - h1 = ```H1(textData: BSTextParams),```
  - h2 = ```H2(textData: BSTextParams),```
  - h3 = ```H3(textData: BSTextParams),```
  - h4 = ```H4(textData: BSTextParams),```
  - h5 = ```H5(textData: BSTextParams),```
  - h6 = ```H6(textData: BSTextParams),```
  ### Selectable Text:
  - p selectable = ```PS(selectData: BSSelectParams),```
  - h1 selectable = ```HS1(selectData: BSSelectParams),```
  - h2 selectable = ```HS2(selectData: BSSelectParams),```
  - h3 selectable = ```HS3(selectData: BSSelectParams),```
  - h4 selectable = ```HS4(selectData: BSSelectParams),```
  - h5 selectable = ```HS5(selectData: BSSelectParams),```
  - h6 selectable = ```HS6(selectData: BSSelectParams),```
  ## Can Set Text Parameters Using BSTextParams:
  ### Regular Text:
```dart
  double textSizeSmall;
  double textSizeLarge;
  String text;
  TextAlign? textAlignment;
  Color? color;
  FontWeight? weight;
  FontStyle? fontStyle;
  TextDecoration? textDecoration;
  double? height;
  Paint? background;
  Color? backgroundColor;
  Color? decorationColor;
  TextDecorationStyle? decorationStyle;
  double? decorationThickness;
  String? fontFamily;
  List<String>? fontFamilyFallback;
  List<FontFeature>? fontFeatures;
  String? debugLabel;
  Paint? foreground;
  bool inherit;
  TextLeadingDistribution? leadingDistribution;
  double? letterSpacing;
  Locale? locale;
  TextOverflow? overflow;
  String? package;
  List<Shadow>? shadows;
  TextBaseline? textBaseline;
  double? wordSpacing;
  int? maxLines;
  String? semanticsLabel;
  bool? softWrap;
  StrutStyle? strutStyle;
  TextDirection? textDirection;
  TextHeightBehavior? textHeightBehavior;
  double? textScaleFactor;
  TextWidthBasis? textWidthBasis;
  Key? textKey;
  double Function(
      BuildContext context, // Build Context used to build text widget
      String textType, { // p, or h1-h6
      required double small, // default BSBreakPoints value unless textSizeSmall is passed
      required double large, // default BSBreakPoints value unless textSizeLarge is passed
  })? customBreakpoints;
```
## Can Set Text Parameters Using BSSelectParams:
### Selectable Text:
``` dart
  String text;
  double textSizeSmall;
  double textSizeLarge;
  void Function()? onTap;
  bool autofocus;
  Color? cursorColor;
  double? cursorHeight;
  Radius? cursorRadius;
  double cursorWidth;
  DragStartBehavior dragStartBehavior;
  bool enableInteractiveSelection;
  FocusNode? focusNode;
  int? maxLines;
  int? minLines;
  void Function(TextSelection, SelectionChangedCause?)? onSelectionChanged;
  ScrollPhysics? scrollPhysics;
  TextSelectionControls? selectionControls;
  BoxHeightStyle selectionHeightStyle;
  BoxWidthStyle selectionWidthStyle;
  String? semanticsLabel;
  bool showCursor;
  StrutStyle? strutStyle;
  TextAlign? textAlign;
  TextDirection? textDirection;
  TextHeightBehavior? textHeightBehavior;
  double? textScaleFactor;
  TextWidthBasis? textWidthBasis;
  ToolbarOptions? toolbarOptions;
  Key? textKey;

  // styles
  Color? color;
  Paint? background;
  Color? backgroundColor;
  String? debugLabel;
  TextDecoration? decoration;
  Color? decorationColor;
  TextDecorationStyle? decorationStyle;
  double? decorationThickness;
  String? fontFamily;
  List<String>? fontFamilyFallback;
  List<FontFeature>? fontFeatures;
  FontStyle? fontStyle;
  FontWeight? fontWeight;
  Paint? foreground;
  double? height;
  bool inherit;
  TextLeadingDistribution? leadingDistribution;
  double? letterSpacing;
  Locale? locale;
  TextOverflow? overflow;
  String? package;
  List<Shadow>? shadows;
  TextBaseline? textBaseline;
  double? wordSpacing;
  double Function(
      BuildContext context, // Build Context used to build text widget
      String textType, { // p, or h1-h6
      required double small, // default BSBreakPoints value unless textSizeSmall is passed
      required double large, // default BSBreakPoints value unless textSizeLarge is passed
  })? customBreakpoints;
```

## Text Sizes:
  Small falls under screen sizes that are defined bootstrap_like_breakpoints for the below, unless ```customBreakpoints``` is passed
  - '',
  - 'sm'

  Large falls under screen sizes that are defined bootstrap_like_breakpoints for the below, unless ```customBreakpoints``` is passed
  - 'md',
  - 'lg',
  - 'xl'
  - 'xxl'

### Sizes in px:
  - p	16px,
  - h1 40px,
  - h1 small 32px,
  - h2 32px,
  - h2 small 25px,
  - h3 28px,
  - h3 small 22px
  - h4 24px,
  - h4 small 24px,
  - h5 20px,
  - h5 small 16px,
  - h6 16px,
  - h6 small 13px

  ## Ref
   - https://bootstrapcreative.com/bootstrap-text-sizes/ 
   - https://pub.dev/packages/bootstrap_like_breakpoints


CHANGELOG.md
## 2.1.1

- Updated license to MIT to add better freedom to the code

## 2.1.0

- added ```customBreakpoints``` function to ```BSTextParams``` and ```BSSelectParams```
    - allows custom logic to determine how to set the size of text to small or large
    - if not passed the default values provided by ```BSBreakPoints.getTextFontSize```

## 2.0.0

- Now can make text selectable, or regular based on data object passed
    - ```BSSelectParams``` for selectable text
    - ```BSTextParams``` for regular text
- consolidated code, removed selectable versions of widgets
- removed ```final``` from fields in params objects
    - this way dynamic changes to text can be performed easier
- removed homepage from yaml

## 1.3.0

- increased tracked files in git repo
- example ```pubspec.yaml``` will now
    - always point to current version of the package
    - not be published
- added ```SelectableText``` versions of params and text
- added ```overflow, locale``` to ```getTextObjectWithFontSizes```
    - the options were missing, but available in ```BSTextParams```
- added key option for text
- removed ```initState```code from all text widgets
    - not needed
- changed ```createState``` return type for all widgets to that widget's state
- added ```dev_dependencies``` and ```uses-material-design: true``` to ```pubspec.yaml``` files
- updated example readme

## 1.2.2

- Added git repo

## 1.2.1

- bootstrap_like_breakpoints: ^0.2.0

## 1.2.0

- Changed from getting breakpoints from bootstrap_like_grid to bootstrap_like_breakpoints
- Changed from having text size values as const values to
    - allowed setting text sizes in BSTextParams
    - default values from bootstrap_like_breakpoints
- Added GNU license to example
- Updated example to no longer require bootstrap_like_grid
- Updated Readme

## 1.1.0

- Changed from using flutter_bootstrap to bootstrap_like_grid
- Updated Readme
- Removed git link, because isn't managed properly yet

## 1.0.4

Added repo, and homepage

## 1.0.3

Properly added example, and added documentation. updated readme. Formatted code.

## 1.0.2

Removed dev_dependencies from YAML. Updated renamed example to main.dart to get more pub points.
Added readme.md to example folder. Updated README. Changed change log so newest changes are at the
top. Removed full fun example from readme. Updated Full example to be cooler.

## 1.0.1

Updated README, included a Full and Fun example. Added the Full and Fun example to
example/full_and_fun_example.dart

## 1.0.0

BootStrapTextData from a required named parameter to a required positional parameter. Updated README
Renamed BootStrapTextData to BSTextParams Rename to bootstrap_typography2 Add the following params
to BSTextParams; height, background, backgroundColor, decorationColor, decorationStyle,
decorationThickness, fontFamily, fontFamilyFallback, fontFeatures, debugLabel, foreground, inherit,
leadingDistribution, letterSpacing, locale, overflow, package, shadows, textBaseline, wordSpacing,
maxLines, semanticsLabel, softWrap, strutStyle, textDirection, textHeightBehavior, textScaleFactor,
textWidthBasis,

## 0.0.2

Fixed 'int is not a subtype of double?' error. updated README

## 0.0.1

Basic version, supports:
supports:
p h1 h2 h3 h4 h5 h6 can control text text alignment color font weight font style text decoration

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:bootstrap_typography2/bootstrap_typography2.dart';

void main() {
  runApp(
    MaterialApp(
      home: Scaffold(
        body: SingleChildScrollView(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              H1(
                textData: BSTextParams("H1"),
              ),
              H2(
                textData: BSTextParams("H2"),
              ),
              H3(
                textData: BSTextParams("H3"),
              ),
              H4(
                textData: BSTextParams("H4"),
              ),
              H5(
                textData: BSTextParams("H5"),
              ),
              H6(
                textData: BSTextParams("H6"),
              ),
              P(
                textData: BSTextParams("P"),
              ),
              H2(
                textData: BSTextParams(
                  "Custom",
                  textSizeSmall: 16,
                  textSizeLarge: 82,
                ),
              ),
              H3(
                textData: BSTextParams(
                  "Custom Breakpoints, and Sizes",
                  textSizeSmall: 16,
                  textSizeLarge: 32,
                  customBreakpoints: (
                    context,
                    textType, {
                    required large,
                    required small,
                  }) {
                    if (MediaQuery.of(context).size.width > 500) {
                      return large;
                    }
                    return small;
                  },
                ),
              ),
              //
              H1(
                selectData: BSSelectParams("H1"),
              ),
              H2(
                selectData: BSSelectParams("H2"),
              ),
              H3(
                selectData: BSSelectParams("H3"),
              ),
              H4(
                selectData: BSSelectParams("H4"),
              ),
              H5(
                selectData: BSSelectParams("H5"),
              ),
              H6(
                selectData: BSSelectParams("H6"),
              ),
              P(
                selectData: BSSelectParams("P"),
              ),
              H2(
                selectData: BSSelectParams(
                  "Custom",
                  textSizeSmall: 16,
                  textSizeLarge: 82,
                ),
              ),
              H3(
                selectData: BSSelectParams(
                  "Custom Breakpoints, and Sizes",
                  textSizeSmall: 16,
                  textSizeLarge: 32,
                  customBreakpoints: (
                    context,
                    textType, {
                    required large,
                    required small,
                  }) {
                    if (MediaQuery.of(context).size.width > 500) {
                      return large;
                    }
                    return small;
                  },
                ),
              ),
            ],
          ),
        ),
      ),
    ),
  );
}
